### PR TITLE
port Base64 Encoder/Decoder

### PR DESCRIPTION
**Pull request by egmkang**
finish Base64 Encoder and Decode and passed the Netty's unittest.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@egmkang__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by nayato** [comment]
@egmkang, I'll look through this today. Sorry about the delay.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, breakLines, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, offset, length, breakLines, dialect, src.Allocator);
+        }
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator alloctor)
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >alloctor [](start = 150, length = 8)

typo

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        {
+            if (src == null)
+            {
+                throw new NullReferenceException("src is null");
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >NullReferenceException [](start = 26, length = 22)

ArgumentNullException

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException("alphabet is null");
+            }
+            if (dialect.alphabet.Length != 64)
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: this can be just Contract.Assert as we make sure dialect is in compliance through testing (no external object is expected here)

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            {
+                throw new Exception(String.Format("alphabet.Length:{0}, not 64!", dialect.alphabet.Length));
+            }
+            if (length <= 0) return Unpooled.Empty;
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: code style:
```
if (length <= 0)
{
    return Unpooled.Empty
}
```

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 1 above)
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >offset [](start = 69, length = 6)

Pls validate offset and length for proper values, especially as you're skipping on range checks internally:
- offset >= 0
- offset + length <= src.ReaderIndex + src.ReadableBytes

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+            int outLength = length / 3 * 4;
+            outLength += length % 3 > 0 ? 4 : 0;
+            outLength += breakLines ? outLength / MAX_LINE_LENGTH : 0;
+            IByteBuffer dest = alloctor.Buffer(outLength).WithOrder(src.Order);
+            //avoid unnecessary range checking
+            fixed (byte* alphabet = dialect.alphabet, srcArray = src.Array, destArray = dest.Array)
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >Array [](start = 69, length = 5)

IByteBuffer.Array may throw if buffer is backed by more than one array. If you want to have a direct access to data in memory that backs the byte buffer, the best approach is to use a combination of `IByteBuffer.IoBufferCount` and then:
- either `IByteBuffer.GetIoBuffer(int, int) if the count is 1.
- or `IByteBuffer.GetIoBuffers(int, int)` if count is > 1.
- or use `IByteBuffer` methods for data management (e.g. GetByte, SetByte) if count is -1 (which means that the buffer's underlying structure cannot be represented as an array).
The "use IByteBuffer methods for data management" approach must be there as a common fallback in any case.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            {
+                int i = 0;
+                int j = 0;
+                int charCount = 0;
+                int calcLength = offset + length - length % 3;
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length % 3 [](start = 51, length = 10)

introduce variable

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+                    {
+                        if (charCount == MAX_LINE_LENGTH)
+                        {
+                            destArray[j++] = NEW_LINE;
+                            charCount = 0;
+                        }
+                        charCount += 4;
+                    }
+
+                    destArray[j + 0] = alphabet[(srcArray[i] & 0xfc) >> 2];
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >srcArray[i] [](start = 49, length = 11)

would it be better to use int* or even long* here and pull all 4 (8) bytes in one go?

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 6 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: is this necessary?

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 1 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: i correct this problem. check offset < src.ReaderIndex && offset + length > src.ReaderIndex + src.ReadableBytes now.

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 5 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: now i check IoBufferCount, when src and dest are all equals to `1`, then using the Pointer's version, else using `Netty`'s GetByte/SetByte version.
and add one simple unittest named TestCompositeBufferDecoder/TestCompositeBufferEncoder.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+                        {
+                            int output = (decodabet[b4[0]] & 0xFF) << 18 |
+                                         (decodabet[b4[1]] & 0xFF) << 12 |
+                                         (decodabet[b4[2]] & 0xFF) << 6;
+                            destArray[charCount++] = (byte)((uint)output >> 16);
+                            destArray[charCount++] = (byte)((uint)output >> 8);
+                        }
+                        else
+                        {
+                            int output = (decodabet[b4[0]] & 0xFF) << 18 |
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >int [](start = 28, length = 3)

why not make it uint?

Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: `>>` must use uint, so the lines below use uint to do the rigth shift.
decodabet is sbyte, so when i change output as uint, the code is similar like this:
```
uint output = (uint)(decodabet[b4[0]] & 0xFF) << 18 |
              (uint)(decodabet[b4[1]] & 0xFF) << 12 |
              (uint)(decodabet[b4[2]] & 0xFF) << 6;
destArray[charCount++] = (byte)(output >> 16);
destArray[charCount++] = (byte)(output >> 8);
```

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 6 above)
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: I just don't trust the compiler with optimizations I guess. Can you guarantee that compiler will optimize introduce variable itself for "length % 3" to avoid calculating it multiple times? If so, then it's ok to leave it as is.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+                //but you can use SlicedByteBuffer to construct one buffer which ReaderIndex and ArrayOffset are both none zero
+                //for example:
+                //var buff = buffer.Slice(1, 100);
+                //buff.ReadInt();
+                //now ArrayOffset is 1 and ReaderIndex is 4
+                //!!!
+                //and the real offset of the array, is ArrayOfffset + offset
+                //!!!
+                int i = src.ArrayOffset + offset;
+                int calcLength = src.ArrayOffset + offset + length - length % 3;
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: I actually meant a variable for `length % 3` specifically.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+            if (offset < src.ReaderIndex || offset + length > src.ReaderIndex + src.ReadableBytes)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            if (length <= 0)
+            {
+                return Unpooled.Empty;
+            }
+
+            int outLength = length / 3 * 4;
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length / 3 * 4 [](start = 28, length = 14)

should this be `length * 4 / 3`?

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+        {
+            return Decode(src, src.ReaderIndex, src.ReadableBytes, dialect);
+        }
+
+        public static IByteBuffer Decode(IByteBuffer src, int offset, int length, Base64Dialect dialect)
+        {
+            return Decode(src, offset, length, dialect, src.Allocator);
+        }
+
+        private unsafe static int DecodeUsingPointer(IByteBuffer src, IByteBuffer dest, sbyte* decodabet, int offset, int length)
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >private [](start = 8, length = 7)

code style: don't use the default visibility scope: ([https://github.com/Azure/DotNetty/wiki/C%23-Coding-Style](https://github.com/Azure/DotNetty/wiki/C%23-Coding-Style), #5)

Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >private [](start = 8, length = 7)

code style: don't spell the default visibility scope ([https://github.com/Azure/DotNetty/wiki/C%23-Coding-Style](https://github.com/Azure/DotNetty/wiki/C%23-Coding-Style), #5)

File: test/DotNetty.Codecs.Tests/Base64Test.cs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        {
+            var src = Unpooled.CopiedBuffer(Encoding.ASCII.GetBytes("____abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcde"));
+            var expectedEncoded = Unpooled.CopiedBuffer(Encoding.ASCII.GetBytes("YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXphYmNkZWZnaGlqa2xtbm9wcXJzdHV2d3h5emFiY2Rl"));
+            var buff = src.Slice(3, src.ReadableBytes - 3);
+            buff.ReadByte();
+            TestEncode(buff, expectedEncoded);
+        }
+
+        [Fact]
+        public void TestSimpleDecode()
```
Comment (nayato) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: >TestSimpleDecode [](start = 20, length = 16)

would be great to cover a variety of test cases, e.g. lengths ranging from 0 to some number enough to wrap the line with NewLine. See [https://github.com/Azure/DotNetty/blob/dev/test/DotNetty.Handlers.Tests/TlsHandlerTest.cs#L63](https://github.com/Azure/DotNetty/blob/dev/test/DotNetty.Handlers.Tests/TlsHandlerTest.cs#L63) for example on how to do this.

**Comment by nayato** [comment]
@egmkang, pls hold off addressing comments, I'll let you know once I'm done reviewing. I'm really sorry for the delay on this.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+        {
+            int charCount = 0;
+            Contract.Assert(dest.WriterIndex == 0 && dest.ArrayOffset == 0);
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >dest.ArrayOffset == 0 [](start = 53, length = 21)

This may not be true. E.g. for PooledByteBuffer retrieved from PooledByteBufferAllocator (most common high perf buffer implementation) ArrayOffset will most likely be non-zero.

File: test/DotNetty.Codecs.Tests/Base64Test.cs [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    using DotNetty.Buffers;
+    using DotNetty.Codecs.Base64;
+    using Xunit;
+
+    public class Base64Test
+    {
+        [Fact]
+        public void TestNotAddNewLineWhenEndOnLimit()
+        {
+            var src = Unpooled.CopiedBuffer(Encoding.ASCII.GetBytes("____abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyzabcde"));
```
Comment (nayato) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: >Unpooled [](start = 22, length = 8)

test with pooled might be interesting as well. E.g. it would've caught an issue with assumption that `dest.ArrayOffset == 0`

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 10 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: ................

File: test/DotNetty.Codecs.Tests/Base64Test.cs (hunk 12 above)
Comment (egmkang) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: `TestEncodeEmpty`, `TestAddNewLine`, `TestDecodeWithLineBreak`, `TestPaddingNewline`
are these not enough?

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 10 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: shoud i write it like this?
`(length / 3) * 4`

i think it is a basic question.

File: test/DotNetty.Codecs.Tests/Base64Test.cs (hunk 12 above)
Comment (nayato) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: I was talking specifically about length of the input and its content. There might be some edge cases there regarding output length calculation, character mapping, etc. Even doing length from 1 to 10 would already cover a lot of grounds in that sense. The sample I referenced does Cartesian on all the input dimensions. Here it might be just one dimension for length or two if input content is to be varied. Considering you're using "fixed" it won't be unreasonable to test it thoroughly

---
In reply to: [76195354](https://github.com/Azure/DotNetty/pull/150#discussion_r76195354) [](ancestors = 76195354)

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 13 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: you mean i allocate a new buffer, and it's ArrayOffset may be non-zero?

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 10 above)
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: I noticed that in netty it is `length * 4 / 3`. Order of operations matter as rounding happens on division, not on the result of the whole expression. E.g. `5 / 3 = 1`, then `1 * 4 = 4`. On the other hand, `5 * 4 = 20`, `20 / 3 = 6`.

---
In reply to: [76196514](https://github.com/Azure/DotNetty/pull/150#discussion_r76196514) [](ancestors = 76196514)

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 13 above)
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: correct. PooledByteBufferAllocator holds continuous byte arrays up to 16 MB and returned pooled byte buffers reference only a portion of it.

---
In reply to: [76196932](https://github.com/Azure/DotNetty/pull/150#discussion_r76196932) [](ancestors = 76196932)

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 9 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]:
```
int remainderLength = length % 3;
int calcLength = src.ArrayOffset + offset + length - remainderLength;
```

File: test/DotNetty.Codecs.Tests/Base64Test.cs (hunk 14 above)
Comment (egmkang) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: now add TestPooledBufferEncode and TestPooledBufferDecode. as you say, PooledBuffer is a little more complex, and the ArrayOffset is more bigger.

File: test/DotNetty.Codecs.Tests/Base64Test.cs (hunk 12 above)
Comment (egmkang) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: now i add two random test named `TestRandomDecode` and `TestRandomEncode`. which generates bytes randomly, and test the decode/encode result with the .NET's `System.Convert`.

Comment (egmkang) [inline test/DotNetty.Codecs.Tests/Base64Test.cs]: english is not my mother language, so it is a little difficult for me to understand what's you are talking about, especially the long text....

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+                int charCount = 0;
+                //commonly, ReaderIndex and ArrayOffset are zero
+                //but you can use SlicedByteBuffer to construct one buffer which ReaderIndex and ArrayOffset are both none zero
+                //for example:
+                //var buff = buffer.Slice(1, 100);
+                //buff.ReadInt();
+                //now ArrayOffset is 1 and ReaderIndex is 4
+                //!!!
+                //and the real offset of the array, is ArrayOfffset + offset
+                //!!!
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: this comment seems misplaced. Looks more appropriate in IByteBuffer.ArrayOffset xml doc in some way

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+                    j += 4;
+                }
+
+                i = calcLength;
+
+                if (breakLines && (length % 3 != 0) && (charCount == MAX_LINE_LENGTH))
+                {
+                    destArray[j++] = NEW_LINE;
+                }
+                switch (length % 3)
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length % 3 [](start = 24, length = 10)

`remainderLength`

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+                    destArray[j + 0] = alphabet[(srcArray[i] & 0xfc) >> 2];
+                    destArray[j + 1] = alphabet[((srcArray[i] & 0x03) << 4) | ((srcArray[i + 1] & 0xf0) >> 4)];
+                    destArray[j + 2] = alphabet[((srcArray[i + 1] & 0x0f) << 2) | ((srcArray[i + 2] & 0xc0) >> 6)];
+                    destArray[j + 3] = alphabet[(srcArray[i + 2] & 0x3f)];
+                    j += 4;
+                }
+
+                i = calcLength;
+
+                if (breakLines && (length % 3 != 0) && (charCount == MAX_LINE_LENGTH))
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length % 3 [](start = 35, length = 10)

`remainderLength`

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+        }
+        private unsafe static int EncodeUsingGetSet(byte* alphabet, IByteBuffer src, IByteBuffer dest, int offset, int length, bool breakLines)
+        {
+            int srcReaderIndex = src.ReaderIndex;
+            int srcWriterIndex = src.WriterIndex;
+
+            int i = 0;
+            int j = 0;
+            int charCount = 0;
+            int calcLength = offset + length - length % 3;
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length % 3 [](start = 47, length = 10)

`remainderLength`

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+                dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                dest.SetByte(j + 1, alphabet[((src.GetByte(i) & 0x03) << 4) | ((src.GetByte(i + 1) & 0xf0) >> 4)]);
+                dest.SetByte(j + 2, alphabet[((src.GetByte(i + 1) & 0x0f) << 2) | ((src.GetByte(i + 2) & 0xc0) >> 6)]);
+                dest.SetByte(j + 3, alphabet[(src.GetByte(i + 2) & 0x3f)]);
+                j += 4;
+            }
+
+            i = calcLength;
+
+            if (breakLines && (length % 3 != 0) && (charCount == MAX_LINE_LENGTH))
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length % 3 [](start = 31, length = 10)

remainderLength

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+            {
+                dest.SetByte(j++, NEW_LINE);
+            }
+            switch (length % 3)
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >length % 3 [](start = 20, length = 10)

`remainderLength`

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 21]
```diff
@@ -0,0 +1,379 @@
...
+                    dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                    dest.SetByte(j + 1, alphabet[(src.GetByte(i) & 0x03) << 4]);
+                    dest.SetByte(j + 2, EQUALS_SIGN);
+                    dest.SetByte(j + 3, EQUALS_SIGN);
+                    j += 4;
+                    break;
+            }
+            //restore src's ReaderIndex and WriterIndex
+            src.SetIndex(srcReaderIndex, srcWriterIndex);
+            //remove last byte if it's NewLine
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: why is this needed? GetByte and SetByte do not change reader and writer indexes.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+                {
+                    if (charCount == MAX_LINE_LENGTH)
+                    {
+                        dest.SetByte(j++, NEW_LINE);
+                        charCount = 0;
+                    }
+                    charCount += 4;
+                }
+
+                dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >src.GetByte(i) [](start = 46, length = 14)

pls introduce variables for GetByte(i). .GetByte(i + 1), GetByte(i + 2) calls to avoid unnecessary duplicate access (compiler likely won't do magic here).

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+            i = calcLength;
+
+            if (breakLines && (length % 3 != 0) && (charCount == MAX_LINE_LENGTH))
+            {
+                dest.SetByte(j++, NEW_LINE);
+            }
+            switch (length % 3)
+            {
+                case 2:
+                    dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >GetByte [](start = 54, length = 7)

see comment above on GetByte

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 24]
```diff
@@ -0,0 +1,379 @@
...
+
+                    b4Count = 0;
+                    if (value == EQUALS_SIGN)
+                    {
+                        break;
+                    }
+                }
+            }
+            //restore src's ReaderIndex and WriterIndex
+            src.SetIndex(srcReaderIndex, srcWriterIndex);
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: GetByte/SetByte methods don't affect reader and writer indexes

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 25]
```diff
@@ -0,0 +1,379 @@
...
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator allocator)
+        {
+            if (src == null)
+            {
+                throw new ArgumentNullException(nameof(src));
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException(nameof(dialect.alphabet));
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >NullReferenceException [](start = 26, length = 22)

ArgumentNullException here and in other places where we check for "not null" and throw if it isn't so.

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 26]
```diff
@@ -0,0 +1,379 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs.Base64
+{
+    using System;
+    using System.Diagnostics.Contracts;
+    using DotNetty.Buffers;
+
+    public class Base64
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >class [](start = 11, length = 5)

make static

File: src/DotNetty.Codecs/Base64/Base64.cs [hunk 27]
```diff
@@ -0,0 +1,379 @@
...
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            Contract.Assert(dialect.decodabet.Length == 127, "decodabet.Length must be 127!");
+            if (length <= 0)
+            {
+                return Unpooled.Empty;
+            }        
+
+            int outLength = length * 3 / 4;
+            IByteBuffer dest = alloctor.Buffer(outLength).WithOrder(src.Order);
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64.cs]: >alloctor [](start = 31, length = 8)

`allocator`

File: src/DotNetty.Codecs/Base64/Base64Encoder.cs [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs.Base64
+{
+    using System;
+    using System.Collections.Generic;
+    using DotNetty.Buffers;
+    using DotNetty.Transport.Channels;
+
+    public class Base64Encoder : MessageToMessageDecoder<IByteBuffer>
```
Comment (nayato) [inline src/DotNetty.Codecs/Base64/Base64Encoder.cs]: >Base64Encoder [](start = 17, length = 13)

mark as sealed. Also please override `Shareable` and return `true` as the encoder itself has no transient state and can be shared across channels

**Comment by nayato** [comment]
ok, I think I'm done with the review. Again, sorry for taking so long to go over it.

File: src/DotNetty.Codecs/Base64/Base64.cs (hunk 10 above)
Comment (egmkang) [inline src/DotNetty.Codecs/Base64/Base64.cs]: i realize what you r talking about.  netty use a temp variable named `e` , which is in Base64.java and line 128(netty 4.1 release).
netty's impl means it allocate a little bigger more space, more than what it really need.
in fact, i read both .net framework and netty's base64 encoding/decoding. i think .net's encoding is  easy to understand, and its decoding is more complex and difficult to understand.

https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/mscorlib/src/System/Convert.cs#L2341

**Comment by nayato** [comment]
Looked through once again and looks good to me. Thanks!
